WooCommerce/Classes/ViewRelated/Products/Edit Product/Composite Product Components/ComponentSettingsViewModel.swift

```diff
@@ -86,13 +124,105 @@ final class ComponentSettingsViewModel: ObservableObject {
+                             storageManager: storageManager)
+    }
+}
+
+// MARK: Private helpers
+private extension ComponentSettingsViewModel {
+    /// Load the component options (products or categories) if needed.
+    ///
+    /// Products are synced from remote, while categories are fetched from storage because they are already fully synced.
+    ///
+    func loadComponentOptions(siteID: Int64,
+                              optionIDs: [Int64],
+                              defaultOptionID: Int64?,
+                              type: CompositeComponentOptionType,
+                              stores: StoresManager,
+                              storageManager: StorageManagerType) {
+        syncProducts(siteID: siteID, optionIDs: optionIDs, defaultOptionID: defaultOptionID, type: type, stores: stores)
+        fetchCategoriesIfNeeded(siteID: siteID, optionIDs: optionIDs, type: type, storageManager: storageManager)
+    }
+
+    /// Syncs products and sets the default option and (if the component options are products) the options list.
+    ///
+    func syncProducts(siteID: Int64, optionIDs: [Int64], defaultOptionID: Int64?, type: CompositeComponentOptionType, stores: StoresManager) {
+        guard productsState == .notLoaded else { return }
+
+        // If the component options are products, sync the products matching the list of option IDs.
+        // Otherwise, the component options are categories, so we only need to sync the default option (a product).
+        let productsToSync = type == .productIDs ? optionIDs : [defaultOptionID].compactMap { $0 }
+        let productAction = ProductAction.retrieveProducts(siteID: siteID, productIDs: productsToSync) { [weak self] result in
+            guard let self else { return }
+
+            switch result {
+            case .success((let products, _)):
+                self.defaultOptionTitle = products.first(where: { $0.productID == defaultOptionID })?.name ?? Localization.noDefaultOption
+                if type == .productIDs {
+                    self.options = products.map { product in
+                        ComponentOption(id: product.productID, title: product.name, imageURL: product.imageURL)
+                    }
+                }
+                self.productsState = .loaded
+            case .failure(let error):
+                if type == .productIDs {
+                    self.options = []
+                }
+                self.productsState = .notLoaded
+                DDLogError("⛔️ Unable to fetch products for composite component settings: \(error)")
+                // TODO-8956: Display notice about loading error
+            }
+        }
+
+        productsState = .loading
+        stores.dispatch(productAction)
+    }
+
+    /// Fetches the categories from storage if the component options are categories.
+    ///
+    func fetchCategoriesIfNeeded(siteID: Int64, optionIDs: [Int64], type: CompositeComponentOptionType, storageManager: StorageManagerType) {
```

Comment:
Sorry if this is a dumb question but why do we fetch categories from our storage layer, and not from network?
Are we sure that they will be always sync by now? I'm only seeing them sync in from the Categories screen.

Reply:
Ugh, not a dumb question at all. You're right, I mixed up when the sync was happening — some categories can be in storage at this point but the full sync only happens from the Categories screen.

I've switched this to fetch the categories from remote in 62e301e. @Ecarrion could you take another look at these changes?